programs/clearing_house/src/math/margin.rs

```diff
@@ -1,91 +1,260 @@
+    calculate_base_asset_value_and_pnl, calculate_base_asset_value_and_pnl_with_oracle_price,
+    calculate_base_asset_value_with_oracle_price,
 };
 use crate::math_error;
 use crate::state::user::User;
 
 use crate::error::ErrorCode;
 use crate::math::amm::use_oracle_price_for_margin_calculation;
-use crate::math::bank_balance::get_balance_value;
+use crate::math::bank_balance::get_balance_value_and_token_amount;
 use crate::math::casting::cast_to_i128;
 use crate::math::oracle::{get_oracle_status, OracleStatus};
 // use crate::math::repeg;
 use crate::math::slippage::calculate_slippage;
+use crate::state::bank::Bank;
 use crate::state::bank::BankBalanceType;
 use crate::state::bank_map::BankMap;
+use crate::state::market::Market;
 use crate::state::market_map::MarketMap;
+use crate::state::oracle::OraclePriceData;
 use crate::state::oracle_map::OracleMap;
 use crate::state::state::OracleGuardRails;
+use crate::state::user::{MarketPosition, UserBankBalance};
 use crate::validate;
+use num_integer::Roots;
 use solana_program::msg;
+use std::cmp::{max, min};
 use std::ops::Div;
 
-#[derive(Copy, Clone)]
+#[derive(Clone, Copy, PartialEq, Debug)]
 pub enum MarginRequirementType {
     Initial,
     Partial,
     Maintenance,
 }
 
+pub fn calculate_size_markup_liability_weight(
+    size: u128, // AMM_RESERVE_PRECISION
+    imf_factor: u128,
+    liability_weight: u128,
+    precision: u128,
+) -> ClearingHouseResult<u128> {
+    if imf_factor == 0 {
+        return Ok(liability_weight);
+    }
+
+    let size_sqrt = ((size / 1000) + 1).nth_root(2); //1e13 -> 1e10 -> 1e5
+
+    let liability_weight_numer = liability_weight
+        .checked_sub(
+            liability_weight
+                .checked_div(max(1, BANK_IMF_PRECISION / imf_factor))
+                .ok_or_else(math_error!())?,
+        )
+        .ok_or_else(math_error!())?;
+
+    // increases
+    let size_markup_liability_weight = liability_weight_numer
+        .checked_add(
+            size_sqrt // 1e5
+                .checked_mul(imf_factor)
+                .ok_or_else(math_error!())?
+                .checked_div(100_000 * BANK_IMF_PRECISION / precision) // 1e5 * 1e2
+                .ok_or_else(math_error!())?,
+        )
+        .ok_or_else(math_error!())?;
+
+    let max_liability_weight = max(liability_weight, size_markup_liability_weight);
+    Ok(max_liability_weight)
+}
+
+pub fn calculate_size_discount_asset_weight(
+    size: u128, // AMM_RESERVE_PRECISION
+    imf_factor: u128,
+    asset_weight: u128,
+) -> ClearingHouseResult<u128> {
+    if imf_factor == 0 {
+        return Ok(asset_weight);
+    }
+
+    let size_sqrt = (size / 1000).nth_root(2) + 1; //1e13 -> 1e10 -> 1e5
+    let imf_numerator = BANK_IMF_PRECISION + BANK_IMF_PRECISION / 10;
+
+    let size_discount_asset_weight = imf_numerator
+        .checked_mul(BANK_WEIGHT_PRECISION)
+        .ok_or_else(math_error!())?
+        .checked_div(
+            BANK_IMF_PRECISION
```

Comment:
if `BANK_IMF_PRECISION`  is 1e6 and `imf_numerator` is 1e5, doesn't the mess with the precision?

Reply:
its `imf_numerator` is 1e6 (with 1/10th added to precision value, ie 1.1)

Reply:
ah i see

now that i see that, why is it hardcoded to 1.1?